### PR TITLE
fix(lint): useless assignment for type definition

### LINKED ISSUE #342: Unexpected [Lint/UselessAssign] Useless assignment to variable
Following is a code example:

```cr
# 1.cr

require "option_parser"

module Translater
  target_language : String? = nil

  OptionParser.parse do |parser|
    parser.banner = <<-USAGE
Usage: translater <option> content
USAGE

    parser.on(
      "-t TARGET",
      "--target=TARGET",
      "Specify target language, support zh-CN|en for now.
default is translate English to Chinese.
Youdao don't support this option.
"
    ) do |target|
      case target
      when "zh-CN"
        target_language = "Chinese"
      when "en"
        target_language = "English"
      else
        STDERR.puts "Supported options: -t zh-CN|en"
        exit
      end
    end
  end

  p target_language
end
```

When run with `ameba 1.cr`, get following unexpected message:

```
1.cr:21:9
[W] Lint/UselessAssign: Useless assignment to variable `target_language`
> target_language = "Chinese"
  ^-------------^

1.cr:23:9
[W] Lint/UselessAssign: Useless assignment to variable `target_language`
> target_language = "English"
  ^-------------^
```

But, this is wrong, if move all code out of module `Translater`, ameba is fine, but it is same for those two cases.

### PR DESCRIPTION
**Pull request by veelenga**
closes #342

### CODE REVIEW

**Comment by Sija** [comment]
@veelenga Is this GTG?

File: spec/ameba/rule/lint/useless_assign_spec.cr [hunk 1]
```diff
@@ -475,6 +475,32 @@ module Ameba::Rule::Lint
...
+
+            bar do
+              foo = "bar"
+            end
+          end
+
+          puts foo
+        )
+        subject.catch(s).should be_valid
+      end
```
Comment (Sija) [inline spec/ameba/rule/lint/useless_assign_spec.cr:503]:
```suggestion
      it "doesn't report if top level variable defined inside class is referenced" do
        expect_no_issues subject, <<-CRYSTAL
          class A
            foo : String? = "foo"

            puts foo
          end
          CRYSTAL
      end

      it "doesn't report if top level variable assigned inside class and referenced" do
        expect_no_issues subject, <<-CRYSTAL
          class A
            foo : String? = "foo"

            bar do
              foo = "bar"
            end

            puts foo
          end
          CRYSTAL
      end
```

File: src/ameba/ast/scope.cr [hunk 2]
```diff
@@ -126,6 +139,11 @@ module Ameba::AST
         ivariables.find(&.name.== "@#{name}")
     end
 
+    # Returns `true` if type declaration variable is assigned in this scope.
+    def assigns_type_dec?(name)
+      type_dec_variables.find(&.name.== name) || outer_scope.try(&.assigns_type_dec?(name))
```
Comment (Sija) [inline src/ameba/ast/scope.cr:144]:
```suggestion
      !!(type_dec_variables.find(&.name.== name) || outer_scope.try(&.assigns_type_dec?(name)))
```

File: src/ameba/ast/variabling/type_def_variable.cr [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        var.name
+      when Crystal::InstanceVar
+        var.name
+      when Crystal::ClassVar
+        var.name
+      when Crystal::Global
+        var.name
+      else
+        raise "unsupported type declaration var node"
+      end
```
Comment (Sija) [inline src/ameba/ast/variabling/type_def_variable.cr:26]:
```suggestion
      case var = @node.var
      when Crystal::Var, Crystal::InstanceVar, Crystal::ClassVar, Crystal::Global
        var.name
      else
        raise "Unsupported type declaration var node"
      end
```

**Comment by veelenga** [comment]
Not yet done. Still need to add tests. Thanks for the feedback.

File: spec/ameba/rule/lint/useless_assign_spec.cr (hunk 1 above)
Comment (veelenga) [reply spec/ameba/rule/lint/useless_assign_spec.cr:503]: I would still use the old style for consistency in the file. Let's move to new spec matchers in a separate PR

File: spec/ameba/ast/variabling/type_def_variable_spec.cr [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    describe "#name" do
+      it "returns var name" do
+        variable = TypeDecVariable.new(type_dec)
+        variable.name.should eq var.name
+      end
+
+      it "raises if type declaration is incorrect" do
+        type_dec = Crystal::TypeDeclaration.new(declared_type, declared_type)
```
Comment (Sija) [inline spec/ameba/ast/variabling/type_def_variable_spec.cr:23]:
```suggestion
        type_dec = Crystal::TypeDeclaration.new(declared_type, declared_type)

```

File: spec/ameba/ast/variabling/type_def_variable_spec.cr [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        expect_raises(Exception, "unsupported var node type: Crystal::Path") do
```
Comment (Sija) [inline spec/ameba/ast/variabling/type_def_variable_spec.cr:24]:
```suggestion
        expect_raises(Exception, "Unsupported var node type: Crystal::Path") do
```

File: spec/ameba/rule/lint/useless_assign_spec.cr [hunk 6]
```diff
@@ -475,6 +475,32 @@ module Ameba::Rule::Lint
...
         issue.message.should eq "Useless assignment to variable `foo`"
       end
+
+      it "doesn't report if top level variable defined inside class is referenced" do
+        s = Source.new %(
+          class A
+            foo : String? = "foo"
+          end
+
+          puts foo
```
Comment (Sija) [inline spec/ameba/rule/lint/useless_assign_spec.cr:485]: Shouldn't this be put _inside_ the class declaration?

File: spec/ameba/rule/lint/useless_assign_spec.cr [hunk 7]
```diff
@@ -475,6 +475,32 @@ module Ameba::Rule::Lint
...
+        s = Source.new %(
+          class A
+            foo : String? = "foo"
+
+            bar do
+              foo = "bar"
+            end
+          end
+
+          puts foo
```
Comment (Sija) [inline spec/ameba/rule/lint/useless_assign_spec.cr:500]: ditto

File: src/ameba/ast/scope.cr [hunk 8]
```diff
@@ -122,8 +135,13 @@ module Ameba::AST
...
     def assigns_ivar?(name)
-      arguments.find(&.name.== name) &&
-        ivariables.find(&.name.== "@#{name}")
+      arguments.any?(&.name.== name) &&
+        ivariables.any?(&.name.== "@#{name}")
+    end
+
+    # Returns `true` if type declaration variable is assigned in this scope.
+    def assigns_type_dec?(name)
+      type_dec_variables.any?(&.name.== name) || outer_scope.try(&.assigns_type_dec?(name))
```
Comment (Sija) [inline src/ameba/ast/scope.cr:144]:
```suggestion
      type_dec_variables.any?(&.name.== name) || !!outer_scope.try(&.assigns_type_dec?(name))
```

File: src/ameba/ast/variabling/type_def_variable.cr [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    delegate end_location, to: @node
+    delegate to_s, to: @node
+
+    def initialize(@node)
+    end
+
+    def name
+      var = @node.var
+
+      case var
```
Comment (Sija) [inline src/ameba/ast/variabling/type_def_variable.cr:15]:
```suggestion
      case var = @node.var
```

File: src/ameba/ast/variabling/type_def_variable.cr [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+      when Crystal::Var, Crystal::InstanceVar, Crystal::ClassVar, Crystal::Global
+        var.name
+      else
+        raise "unsupported var node type: #{var.class}"
```
Comment (Sija) [inline src/ameba/ast/variabling/type_def_variable.cr:19]:
```suggestion
        raise "Unsupported var node type: #{var.class}"
```

File: spec/ameba/rule/lint/useless_assign_spec.cr (hunk 6 above)
Comment (veelenga) [reply spec/ameba/rule/lint/useless_assign_spec.cr:485]: Nope, we try to match the original scenario in the opened issue

**Review by Sija — APPROVED** [review]
GTG from my side, yet wait please with the merge until we figure out what to do with current `1.4.0` release.

**Pull request merged**